src/chainrules.jl

```diff
@@ -0,0 +1,63 @@
+@non_differentiable validate_woodbury_arguments(A, D, S)
+
+# Rule for Woodbury * Real. 
+# Ignoring Complex version for now. 
+# NOTE: # Can't use Tangent{WoodburyPDMat} here. ChainRules is fine, but Zygote has issues and passes around Tangent{Any}
+
+function ChainRulesCore.rrule(::typeof(*), A::WoodburyPDMat, B::Real)
+    project_A = ProjectTo(A)
+    project_B = ProjectTo(B)
+    primal = A * B
+    times_pullback(ȳ) = _times_pullback(ȳ, primal, A, B, (;A=project_A, B=project_B))
+    return primal, times_pullback
+end
+
+function ChainRulesCore.rrule(::typeof(*), A::Real, B::WoodburyPDMat, )
+    project_A = ProjectTo(A)
+    project_B = ProjectTo(B)
+    primal = A * B
+    times_pullback(ȳ) = _times_pullback(ȳ, primal, A, B, (;A=project_A, B=project_B))
+    return primal, times_pullback
+end
+
+_times_pullback(ȳ::AbstractThunk, primal, A, B, proj) = _times_pullback(unthunk(ȳ), primal, A, B, proj)
+# If the cotangent is a Matrix we first need to project down, otherwise ignore
+_times_pullback(Ȳ::AbstractMatrix, primal, A, B, proj) = _times_pullback(ProjectTo(primal)(Ȳ), A, B, proj)
+_times_pullback(ȳ::Tangent, primal, A, B, proj) = _times_pullback(ȳ, A, B, proj)
+
+function _times_pullback(Ȳ::Tangent, A::T, B::Real, proj)  where {T<:WoodburyPDMat}
+    Ā = @thunk proj.A(Tangent{WoodburyPDMat}(; A = Ȳ.A, D = Ȳ.D * B', S = Ȳ.S * B'))
+    B̄ = @thunk proj.B(dot(Ȳ.D, A.D) + dot(Ȳ.S, A.S))
+    return (NoTangent(), Ā, B̄)
+end
+
+function _times_pullback(Ȳ::Tangent, A::Real, B::T, proj) where {T<:WoodburyPDMat}
+    Ā = @thunk proj.A(dot(Ȳ.D, B.D) + dot(Ȳ.S, B.S))
+    B̄ = @thunk proj.B(Tangent{WoodburyPDMat}(; A = Ȳ.A, D = Ȳ.D * A, S = Ȳ.S * A))
+    return (NoTangent(), Ā, B̄)
+end
+
+# Composite pullbacks
+function ChainRulesCore.rrule(
+    ::Type{T},
+    A::AbstractMatrix,
+    D::Diagonal,
+    S::Diagonal,
+    ) where  {T<:WoodburyPDMat}
+    return WoodburyPDMat(A, D, S), X̄ -> WoodburyPDMat_pullback(X̄, A, D, S)
+end
+WoodburyPDMat_pullback(X̄::Tangent, A, D, S) = (X̄ = (unthunk(X̄); return (NoTangent(), X̄.A, X̄.D, X̄.S)))
+WoodburyPDMat_pullback(X̄::AbstractThunk, A, D, S) = WoodburyPDMat_pullback(unthunk(X̄), A, D, S)
+
+function ChainRulesCore.ProjectTo(W::T) where {T<:WoodburyPDMat}
+    fields = (A = W.A, D = W.D, S = W.S)
+    ChainRulesCore.ProjectTo{T}(; fields...)
+end
+
+function (project::ProjectTo{T})(X̄::AbstractMatrix) where {T<:WoodburyPDMat}
```

Comment:
I don't understand what this does. It looks like we are trying to project an arbitrary matrix onto a Woodbury, which as I understand is not possible in case X is not positive semi-definite? (In that case we might explicitly check for PSD and throw an error?)

If all is well and I just don't understand something, could we add a few lines of comments that explain what is going on and maybe link a reference for the math?

Reply:
> trying to project an arbitrary matrix onto a Woodbury

Nope - we are projecting an arbitrary matrix onto a Woodbury _tangent_. I.e. instead of passing around W̄
, we are passing around Ā, D̄, S̄

The matrix here can be anything. Project here is encoding the rules on how the individual components of the woodbury, i.e. from the definition `W = ADA' + S -> dW = ADdA' + AdDA' + dS` => Ā = 2ADW̄, D̄=AW̄A', S̄ = W̄ with their appropriate projectiosn down  to e.g. Diagonal. X̄ here is W̄.

Ideally, these would only be in the constructor for the Woodbury, however, due to the way FD is set-up, it seems we always have to project down to these components. THe woodbury is a lazy low-rank matrix and because of this any woodbury primal has number of elements 2*(A1 + A2) where (A1, A2) is the size of the matrix A. And the tangent has to be the same size (i.e. number of elements), i.e. their `to_vec` representation must equal in length

This is why when testing with a Matrix cotangent I couldn't use CRTU and had to do a manual call to FD and densify the primal [here](https://github.com/invenia/PDMatsExtras.jl/blob/ff8d903bba668e8be6a115866e6abfc1a12d4700/test/chainrules.jl#L35)

If need be, I believe we  can only ever return the natural differential Matrix (i.e. W̄ ) as the outut of any abitrary rrule with a Woodbury argument as is done [here](https://github.com/invenia/PDMatsExtras.jl/blob/ar/chainrules/src/chainrules.jl#L25) but as I say I believe that FD will not like this in some scenarios. 

The constraints imposed here are, as I understand them are that the primal output and output cotangent need to be of the same length, and that the input args and input adjoints are also all of the same length to each other (where length is the to_vec representation). 

I'll add comments, but first I want to make sure that that makes sense

Reply:
Hmm, I am actually not sure what we should do here, I've opened https://github.com/JuliaDiff/ChainRulesCore.jl/issues/442
with the `PositiveReal` example you have showed me.

Reply:
I'm reading that thread. 

The use case here aside, Is projecting onto a `Tangent` not an intended use of `ProjectTo`?
I'm sort of getting that impression from reading these threads, and it is not one that I had realised, if that is the case.

Reply:
Sure, we can map to `Tangent` because `Tangent` is a differential type. The differences is that
```julia
ProjectTo(w::Tangent{Woodbury})
```
is fine, but
```julia
ProjectTo(w::Woodbury)
```
is not, and we are doing the second here. (Despite the fact that the projection actually returns the `Tangent{Woodbury}`, which is inconsistent since we said we are projecting onto a `Woodbury`)

Reply:
Ohhhhhh I see. I'd misunderstood the semantics of `ProjectTo` (and the objection here). 

In which case, I think what I want to do is impossible under the set-up here as it doesn't make sense to keep the woodbury, or in fact any constrained type. 

So any rrule with ProjectTo need to have custom rules when it uses constrained types

But I think this conversation can now fall under the issue you created.

Reply:
yeah let's see what conclusion we reach there. (I suspect that we will allow this, but let's not merge this until we actually do)

Reply:
How does this relate to the proposal in https://github.com/JuliaDiff/ChainRulesCore.jl/pull/449?

Reply:
I think Will thought that it is possible to find the inverse of Woodbury densification (for the subspace of dense matrices that _can_ come from a Woodbury) but I may have misinterpreted that.

This seems to be a weird edge case @willtebbutt. If I understand correctly, in this case we want the structural rather than the natural differential inside the pullback written in the rule (rather than the premise in the 449 which expects AD to prefer structural tangents, and rule writers to prefer natural tangents)

Reply:
The proposal in 449 would basically replace `ProjectTo` / change its semantics to eat a natural tangent / `Matrix` and produce a structural tangent. You obtain the new thing by implementing the pullback for `collect`, and that does everything you need.

Re inverse of densification -- I found that hard to implement for `WoodburyPDMat` -- see [here](https://github.com/JuliaDiff/ChainRulesCore.jl/pull/449/files?file-filters%5B%5D=.jl&file-filters%5B%5D=.md#diff-7b46aec105a00326f97d8740cefcee1c791fb89447f6637410d4855b8b84ffe2R475) but the pullback for `collect` (`pullback_of_destructure` -- I wanted to give it a different name) is totally fine.

Reply:
Separately, what's the rationale for implementing `ProjectTo`? Is it to make it possible to do `*(A::Matrix, B::WoodburyPDMat)`? If so, would it make more sense / be easier to implement a specialised method for the primal (since the generic fallback will be horribly slow) that is differentiable, and `@opt_out` of the generic rrule to let AD do the work of deriving the rule?

Reply:
> Separately, what's the rationale for implementing ProjectTo

Originally, I was using the `ProjectTo` here to project the natural tangent onto it's structural tangent - i.e. for a given W̄ (the natural tangent, as I understand the semantics), to it's structural (Ā, D̄, S̄), but it turned out this wasn't permitted under the semantics of ProjectTo - which what led to the 'allow projectto onto tangent' [issue ](https://github.com/JuliaDiff/ChainRulesCore.jl/issues/442). I paused looking at this MR at that point, but saw your RFC being posted and pinged this thread on how these two related. As far as I understand the resolution to #442 in CRC it's added an identity, but I don't think that changes anything w.r.t using it as a tangent, so I still wouldn't know how to continue this. 

*If* I understand your RFC, the way I was using the projectto here was more in line with what you were considering.